### PR TITLE
Add library for P2P communication and sample app

### PR DESCRIPTION
**Pull request by espertus**
Add library for P2P communication and sample app

This adds `lib-nearby` on top of the Google Play Nearby API.
The sample app `nearby-chat` uses it to create a chat application enabling
communication between 2 devices.

This is a better version of https://github.com/mozilla-mobile/android-components/pull/4794 because it is a single commit.

---

### Pull Request checklist

- [X] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [X] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [X] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -138,6 +139,9 @@ permalink: /changelog/
   * Deprecated `PromptFeature` constructor that has parameters for both `Activity` and `Fragment`. Use the constructors that just take either one instead.
   * Changed `sessionId` parameter name to `customTabId`.
 
+* **sample-nearby-chat**
+  * 🆕 New sample program demonstrating use of `lib-nearby`.
+
```
Comment (jonalmeida) [inline docs/changelog.md]: These entries should probably go together in the latest release version.

**Comment by jonalmeida** [comment]
I filed https://github.com/mozilla-mobile/android-components/issues/4984 for the code coverage report as a follow-up.

**Comment by espertus** [comment]
Thank you for letting me defer those.

**Comment by espertus** [comment]
This works for me: `./gradlew :samples-nearby-chat:assembleDebug`

I'll try the other builds run by CI.

**Comment by espertus** [comment]
I found the failing task: `:samples-nearby-chat:test`

I'll see what's going on.

File: components/feature/readerview/src/main/assets/extensions/readerview/readerview.js [hunk 2]
```diff
@@ -70,7 +70,7 @@ class ReaderView {
     result.language = document.documentElement.lang;
 
     let article = Object.assign(
-      result,
+      result,r
```
Comment (jonalmeida) [inline components/feature/readerview/src/main/assets/extensions/readerview/readerview.js]: It looks like this is an artifact from the other git merge issues. :)

File: samples/sync/build.gradle [hunk 3]
```diff
@@ -45,4 +45,6 @@ dependencies {
 
     implementation Dependencies.kotlin_reflect
     implementation Dependencies.androidx_recyclerview
+    implementation Dependencies.androidx_appcompat
```
Comment (jonalmeida) [inline samples/sync/build.gradle]: More artifacts from the merge issue.

File: components/feature/readerview/src/main/assets/extensions/readerview/readerview.js (hunk 2 above)
Comment (espertus) [reply components/feature/readerview/src/main/assets/extensions/readerview/readerview.js]: I'm not sure how that got in there. Sorry!

File: docs/changelog.md (hunk 1 above)
Comment (espertus) [reply docs/changelog.md]: Absolutely. I also fixed a typo on an earlier entry  when fixing mine.

**Comment by espertus** [comment]
I added a big fat "unit" test but suspect you'd like for me to pull it apart into lots of little tests, which I can do if I create a setter for `NearbyConnection.connectionState` for use from tests (yuck). Let me know which option you prefer (big fat test or injecting `ConnectionState`s).

I still don't know how to fix the build problem but will give it another try.

**Comment by jonalmeida** [comment]
One big test or many small ones, as long as it's readable/maintainable and increases code coverage. 🙂

**Comment by espertus** [comment]
The tests were written, a bug was fixed, and the commits were squashed. If all the checks succeed and you approve of the changes, I think we should be set.

**Comment by espertus** [comment]
Test coverage is now 86% (according to my client).

**Comment by espertus** [comment]
GitHub says there are outstanding requests for changes, but I'm having trouble finding them. I'd love to get this checked in.

**Comment by jonalmeida** [comment]
bors r=jonalmeida

@espertus All green! ✅  ✅  ✅

**Comment by pocmo** [comment]
bors retry

**Comment by pocmo** [comment]
bors retry

**Pull request merged**